pydicom/valuerep.py

```diff
@@ -350,28 +490,70 @@ def __init__(
         elif isinstance(val, (DSfloat, DSdecimal)) and has_attribute:
             self.original_string = val.original_string
 
+        self.auto_format = auto_format
+        if self.auto_format:
+            # If auto_format is True, keep the float value the same, but change
+            # the string representation stored in original_string if necessary
+            if hasattr(self, 'original_string'):
+                if not is_valid_ds(self.original_string):
+                    self.original_string = format_number_as_ds(
+                        float(self.original_string)
+                    )
+            else:
+                self.original_string = format_number_as_ds(self)
+
+        if config.enforce_valid_values:
+            if len(repr(self).strip('"')) > 16:
+                raise OverflowError(
+                    "Values for elements with a VR of 'DS' must be <= 16 "
+                    "characters long, but the float provided requires > 16 "
+                    "characters to be accurately represented. Use a smaller "
+                    "string, set 'config.enforce_valid_values' to False to "
+                    "override the length check, or explicitly construct a DS "
+                    "object with 'auto_format' set to True"
+                )
+            if not is_valid_ds(repr(self).strip('"')):
+                # This will catch nan and inf
```

Comment:
More or less repeating the `is_valid_ds` call here if both `auto_format` and `enforce_valid_values` are True, which heightens the concern about performance in my other comment.  Perhaps there is a way to avoid the double call, by adjusting the branching logic?

Reply:
Yes, in fact there's no reason to run the test again if `auto_format` is `True` because in that case one of the following will already have happened:
- The user passed a valid string that was already tested
- The user passed an invalid string that was already formatted correctly.

I guess I left this in before out of an abundance of caution, but it should be removed. I removed in d8f8424